Airtime Billing page - Add support for August promotion plans

drigato opened the pull request.

drigato committed:
SAAS-973: Airtime Billing page - Add support for August promotion plans

created promo-eligibility-check endpoint

drigato committed:
SAAS-973: Airtime Billing page - Add support for August promotion plans

Backend side pretty much done

drigato committed:
SAAS-973: Airtime Billing page - Add support for August promotion plans

Un-hardcode product ids

asantoni commented on airtime_mvc/application/common/Billing.php:
```diff
@@ -326,4 +328,123 @@ public static function addVatToInvoice($invoice_id)
+        $postfields["action"] = "getproducts";
+        $postfields["responsetype"] = "json";
+        //gid is the Airtime product group id on whmcs
+        $postfields["gid"] = "15";
```
Can you make this GID a constant as well?

asantoni commented on airtime_mvc/application/common/Billing.php:
```diff
@@ -326,4 +328,123 @@ public static function addVatToInvoice($invoice_id)
+        }
+
+        // if client is currently on monthly or old plan AND (upgrading OR  upgrading/downgrading to annual plan), YES
+        if ($currentPlanBillingCycle == "monthly" || $currentPlanBillingCycle == "free account") {
```
is "free account" the correct string returned by WHMCS?

drigato replied:
yeah it is, i checked and tested it

asantoni commented on airtime_mvc/application/views/scripts/billing/upgrade.phtml:
```diff
@@ -274,6 +298,9 @@ echo($currentProduct["name"]);
     Save 15% on annual plans (Hobbyist plan excluded).
     </div>
+     <div id="promo-plan-eligible" style="display:none;">
+         Congratulations, you are eligible for our Awesome Promotional Plan!
```
Should we say "an" instead of "our"?? (there are multiple promo plans)

asantoni commented on airtime_mvc/application/views/scripts/billing/upgrade.phtml:
```diff
@@ -120,17 +120,41 @@ function configureByCountry(countryCode)
+    var newproductbillingcycle = $("input[type='radio'][name='newproductbillingcycle']:checked").val();
+
+    $.post("/billing/promo-eligibility-check", {"newproductid": newproductid,
+        "newproductbillingcycle": newproductbillingcycle})
```
This AJAX POST needs to pass the CSRF token along too.

asantoni commented on airtime_mvc/application/controllers/BillingController.php:
```diff
@@ -19,6 +20,24 @@ public function indexAction()
+        $request = $this->getRequest();
+        if (!$request->isPost()) {
+            throw new Exception("Must POST data to promoEligibilityCheckAction.");
+        }
```
You should check for the CSRF token here. 

Check out uploadAction() in PluploadController.php for an example...

asantoni replied:
On the PHP side, in the billing form, add:

$csrf_namespace = new Zend_Session_Namespace('csrf_namespace');
$csrf_element = new Zend_Form_Element_Hidden('csrf');
$csrf_element->setValue($csrf_namespace->authtoken)->setRequired('true')->removeDecorator('HtmlTag')->removeDecorator('Label');
$form->addElement($csrf_element);

In the template, you can put:
$this->form->getElement('csrf')

In Javascript, you can add this to the dictionary: 
    "csrf_token" : $("#csrf").attr('value')

drigato committed:
SAAS-973: Airtime Billing page - Add support for August promotion plans

Front-end changes to price grid, and plan title

drigato committed:
SAAS-973: Airtime Billing page - Add support for August promotion plans

Made WHMCS Airtime group id a constant
Check for CSRF token on promo eligibilty ajax check

drigato committed:
Merge branch 'saas' into saas-aug-promo

drigato committed:
SAAS-973: Airtime Billing page - Add support for August promotion plans

Fix to not allow current promo plans to downgrade to another promo plan
Added promo banner to billing page

drigato committed:
Forgot to commit promo banner

drigato committed:
Fix for bringing customer to view invoice that doesn't exist after downgrade

drigato committed:
Show form on billing page if no invoice was generated after downgrade

drigato committed:
Un-doing last 2 committs about invoices

drigato committed:
Fix for billing downgrade with no new invoice created
